### ISSUE TITLE
stateful lstm

### ISSUE THREAD

**Issue by blablaen** [fchollet/keras#1870]
Hi developers,

I run over the stateful_lstm.py example and see some problems:
1. The data doesn't arrange as it suppose to be in stateful mode - the batch_size are 25 and "cos" doesn't arrange in like it say Keras FAQ - " If X1 and X2 are successive batches of samples, then X2[i] is the follow-up sequence to X1[i], for every i".
2. When i turn off stateful mode (stateful=False) i get the good results:
stateful=True  >>> loss = ±0.28
stateful=False >>> loss = ±0.3

Beside that, I just built a model using stateful lstm layers, and have some issue, i saw that when i turn on the stateful mode, the train loss getting better but the test loss are getting very bad.
After debugging the data preprocessing step (and see that it has no bugs), i really think that it has some bug in the lstm stateful implementation. (I order the data as it write in Keras FAQ for stateful mode).
When i turn off the stateful mode the model archive better results in the test set.

Has anyone been able to train stateful network that achieve good result compare to unstateful network?

(I use the latest version of Keras)

Thanks!

**Comment by lukovkin** [comment]
It is also discussed here - https://github.com/fchollet/keras/issues/1820.
Some points:
1. Was thinking a lot about it and I probably agree that example contradict with the docs. But I'm not sure which is more correct.
Let's say we have sequence of 50000 steps and batch_size=25 (as in example) and consider 2 cases:

- Like in the docs, X2[i] = X1[i+1], so samples in the batches will overlap, e.g. X1 will contain samples from 1 to 25 (X[1:26]), X2 will contain samplesfrom 2 to 26 (X[2:27]). In this case the state of the stateful layer after the processing the batch X1 will be the function of all previous inputs: S1 = f(X[0:26]). But then we will feed into the layer the inputs from the batch X2 (X[2:26]) that the layer has already 'seen' and encompassed into the state S1. State after the batch X2 will be the function of the state S1, inputs X[2:26] (which are already in state S1) and input X[27], seen for the first time. Then it will repeat with each subsequent batch. I don't know if it is right or not.

- Like in example X2[0] = X1[end+1] (not correct, but for illustration). So samples do not overlap, X1 = X[1:26], X2 = X[26:51], etc. In this case, state S1 will be the function of X[1:26], then during processing of X2, state S1 and inputs X[26:51] will be taken into account. So state S2 will be the function of S1 and inputs X[26:51]. Make sense too, no?

So it is not 100% clear for me.

**Comment by blablaen** [comment]
Hi,
Thank you for the comment.

I think that the second case (samples do not overlap, X1 = X[1:26], X2 = X[26:51]) you wrote is the good one. But the Keras example stateful_lstm.py doesn't arrange the data in this way:

- the batch_size = 25
- X1[1] = cos at time 1.
- X2[1] = cos at time 26.

whereas it should be - X2[1] = cos at time 2. Did i miss something?

**Comment by lukovkin** [comment]
Er...
Vice versa, I think.
Second case (samples do not overlap, X1 = X[1:26], X2 = X[26:51]) is exactly what we see in example: X2[0] = cos[26] (let's stick to the zero-starting ;).
May be I'm wrong, but it really seems like this.

**Comment by blablaen** [comment]
I just look at X (which is cos in our example):
X[0] = cos[0]
X[1] = cos[1]
X[2] = cos[2]
X[3] = cos[3]
X[4] = cos[4]
.
.
.
X[n] = cos[n]

so, the problem is that X1[0] = cos[0] wheras X2[0] = cos[26], when it should be X2[0]=cos[1],
like it wrote in the FAQ - 
" If X1 and X2 are successive batches of samples, then X2[i] is the follow-up sequence to X1[i], for every i"

So i think it's a problem in the example. Am i right?

**Comment by santi-pdp** [comment]
Something confusing for me in the FAQ example is that the format is still the same as when there was no stateful option, so we have (samples,timesteps,input_dim). I thought we could feed, as in cosine example, the batches like (batch_size,1,input_dim), where batch_size does now the "timestep function" as it does not get reset after every batch. What's the point of feeding the truncated maxlen sequence?

**Comment by blablaen** [comment]
OK, i mange to train stateful lstm network. 
something i have notice is that - the model doesn't reset his state before he run the validation set which cause high loss value compare to the training set.

**Comment by philipperemy** [comment]
Have a look at this guys. Might help you understand how stateful lstm work http://philipperemy.github.io/keras-stateful-lstm/
